File: go/pkg/files/files_1.go
package files

import (
	"fmt"
	"io/ioutil"
	"strings"
)

// ReadFile parses the file and creates
// a map of words to their indices
func ReadFile(path string) (map[string][]int, error) {
	bs, err := ioutil.ReadFile(path)

	if err != nil {
		return nil, err
	}

	m := map[string][]int{}

	for i, w := range strings.Fields(string(bs)) {
		m[w] = append(m[w], i+1)
	}

	return m, nil
}

// absolute value
func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

// minimum
func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

// O(m + n) find the element in idx1 and the element in idx2
// that are closest to each other and return the distance
func closestPair(idx1, idx2 []int) int {
	if len(idx1) == 1 && len(idx2) == 1 {
		return abs(idx1[0] - idx2[0])
	}

	if len(idx1) == 1 {
		distance := 0
		currentDistance := 0
		i := 0

		for i < len(idx2) && (currentDistance == 0 || distance < currentDistance) {
			currentDistance = distance
			distance = abs(idx1[0] - idx2[i])
			i++
		}

		return min(distance, currentDistance)
	} else if len(idx2) == 1 {
		distance := 0
		currentDistance := 0
		i := 0

		for i < len(idx1) && (currentDistance == 0 || distance < currentDistance) {
			currentDistance = distance
			distance = abs(idx2[0] - idx1[i])
			i++
		}

		return min(distance, currentDistance)
	}

	i := 0
	j := 0
	minDistance := -1
	distance := 0

	for i < len(idx1)-1 && j < len(idx2)-1 {
		distance = abs(idx1[i] - idx2[j])
		iDistance := abs(idx1[i+1] - idx2[j])
		jDistance := abs(idx1[i] - idx2[j+1])

		if minDistance == -1 || distance < minDistance {
			minDistance = distance
		}

		if iDistance < jDistance {
			i++
		} else {
			j++
		}
	}

	currentDistance := distance

	if i == len(idx1)-1 {
		for j < len(idx2) && distance <= currentDistance {
			j++
			currentDistance = distance
			distance = abs(idx1[i] - idx2[j])
		}
	} else if i != len(idx1)-1 {
		for i < len(idx1) && distance <= currentDistance {
			i++
			currentDistance = distance
			distance = abs(idx1[i] - idx2[j])
		}
	}

	return min(minDistance, currentDistance)
}

// MinWordDistance finds the minimum distance (number of words)
// between two words in a text file
func MinWordDistance(m map[string][]int, w1, w2 string) (int, error) {
	idx1, ok1 := m[w1]
	idx2, ok2 := m[w2]

	if !ok1 && !ok2 {
		return 0, fmt.Errorf("neither %s or %s appear in the file", w1, w2)
	}
	if !ok1 {
		return 0, fmt.Errorf("%s does not appear in the file", w1)
	}
	if !ok2 {
		return 0, fmt.Errorf("%s does not appear in the file", w2)
	}

	return closestPair(idx1, idx2), nil
}

File: python/files/files_1.py
# Approach:
# read the file and create a dictionary
# assuming we can hold this dictionary in memory
# where each word maps to an ordered list of indices
# such that an entry "foo": [1, 123, 1928] denotes that
# the word "foo" is the 1st, 123rd, and 1928th word in the file

# then we can restate the problem as
# given two lists of integers U and V
# determine the minimum absolute difference
# between any element of U and any element of V
# | u_i - v_j | for all u_i in U and v_j in V

# maintain an index pointer for each ordered list initialized to the start
# set the current minimum distance to the absolute difference of the first elements
# increment the index pointer that results in the smaller absolute distance
# note, this may be worse than the current minimum
# keep track of the minimum distance encountered across all iterations
# eventually, one pointer will reach the end of the list
# advance the other pointer so long as current distance is decreasing
# once distance starts increasing it will never get smaller
# since one list has been completely exhausted and the lists are ordered
# return the minimum distance encountered in O(m + n)

from collections import defaultdict


def read_file():
    word_to_indices = defaultdict(list)
    i = 1

    with open('file_1.txt') as f:
        # lazy for a multi-line file
        # in the event of a single-line file Python can still handle
        # files that exceed available memory by using swap
        for line in f:
            words = line.split()

            for j, word in enumerate(words):
                word_to_indices[word].append(i + j)

            i += len(words)

    return word_to_indices


# O(m + n) find the element in arr1 and the element in arr2
# that are closest to each other and return the distance
def closest_pair(arr1, arr2):
    if len(arr1) == 1 and len(arr2) == 1:
        return abs(arr1[0] - arr2[0])

    # m || n == 1 case could be handled in O(log n) using binary search
    if len(arr1) == 1:
        distance = None
        current_distance = None
        i = 0
        while i < len(arr2) and (current_distance is None or distance < current_distance):
            current_distance = distance
            distance = abs(arr1[0] - arr2[i])
            i += 1
        return min(distance, current_distance)
    elif len(arr2) == 1:
        distance = None
        current_distance = None
        i = 0
        while i < len(arr1) and (current_distance is None or distance < current_distance):
            current_distance = distance
            distance = abs(arr2[0] - arr1[i])
            i += 1
        return min(distance, current_distance)

    i = 0
    j = 0
    min_distance = None
    distance = 0

    while i < len(arr1) - 1 and j < len(arr2) - 1:
        distance = abs(arr1[i] - arr2[j])
        advance_i_distance = abs(arr1[i+1] - arr2[j])
        advance_j_distance = abs(arr1[i] - arr2[j+1])

        if min_distance is None or distance < min_distance:
            min_distance = distance

        if advance_i_distance < advance_j_distance:
            i += 1
        else:
            j += 1

    current_distance = distance

    if i == len(arr1) - 1:
        while j < len(arr2) and distance <= current_distance:
            j += 1
            current_distance = distance
            distance = abs(arr1[i] - arr2[j])
    elif i != len(arr1) - 1:
        while i < len(arr1) and distance <= current_distance:
            i += 1
            current_distance = distance
            distance = abs(arr1[i] - arr2[j])

    return min(min_distance, current_distance)


def min_word_distance(w1, w2):
    word_to_indices = read_file()

    idx1 = word_to_indices.get(w1)
    idx2 = word_to_indices.get(w2)

    if not idx1 and not idx2:
        raise Exception(f'neither {w1} or {w2} appear in the file')
    if not idx1:
        raise Exception(f'{w1} does not appear in the file')
    if not idx2:
        raise Exception(f'{w2} does not appear in the file')

    return closest_pair(idx1, idx2)
